pkg/cache/controller_test.go

```diff
@@ -0,0 +1,110 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"testing"
+
+	"github.com/kcp-dev/logicalcluster"
+	"github.com/stretchr/testify/require"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+func newUnstructured(cluster, namespace, name string, labels labels.Set) *unstructured.Unstructured {
+	u := new(unstructured.Unstructured)
+	u.SetClusterName(cluster)
+	u.SetNamespace(namespace)
+	u.SetName(name)
+	u.SetLabels(labels)
+	return u
+}
+
+func TestClusterIndexFunc(t *testing.T) {
+	tests := map[string]struct {
+		obj     *unstructured.Unstructured
+		desired string
+	}{
+		"bare cluster":             {obj: newUnstructured("test", "", "", nil), desired: "test//"},
+		"bare cluster with dashes": {obj: newUnstructured("test-with-dashes", "", "", nil), desired: "test-with-dashes//"},
+	}
+	for name, tt := range tests {
+		t.Run(name, func(t *testing.T) {
+			result, err := ClusterIndexFunc(tt.obj)
+			require.NoError(t, err, "unexpected error calling ClusterIndexFunc")
+			require.Len(t, result, 1, "ClusterIndexFunc should return one result")
+			require.Equal(t, result[0], tt.desired)
+
+			clusterName := logicalcluster.From(tt.obj).String()
+			namespace := tt.obj.GetNamespace()
+			name := tt.obj.GetName()
+			key := ToClusterAwareKey(clusterName, namespace, name)
+
+			require.Equal(t, result[0], key, "ClusterIndexFunc and ToClusterAwareKey functions have diverged")
+		})
+	}
+}
+
+func TestClusterAndNamespaceIndexFunc(t *testing.T) {
+	tests := map[string]struct {
+		obj     *unstructured.Unstructured
+		desired string
+	}{
+		"bare cluster":          {obj: newUnstructured("test", "", "", nil), desired: "test//"},
+		"cluster and namespace": {obj: newUnstructured("test", "testing", "", nil), desired: "test/testing/"},
+	}
+	for _, tt := range tests {
+		t.Run(tt.desired, func(t *testing.T) {
+			result, err := ClusterAndNamespaceIndexFunc(tt.obj)
+			require.NoError(t, err, "unexpected error calling ClusterAndNamespaceIndexFunc")
+			require.Len(t, result, 1, "ClusterIndexFunc should return one result")
+			require.Equal(t, result[0], tt.desired)
+
+			clusterName := logicalcluster.From(tt.obj).String()
+			namespace := tt.obj.GetNamespace()
+			name := tt.obj.GetName()
+			key := ToClusterAwareKey(clusterName, namespace, name)
+
+			require.Equal(t, result[0], key, "ClusterAndNamespaceIndexFunc and ToClusterAwareKey functions have diverged")
+		})
+	}
+}
+
+func TestClusterAwareKeyFunc(t *testing.T) {
+	tests := map[string]struct {
+		obj     *unstructured.Unstructured
+		desired string
+	}{
+		"cluster and namespace":       {obj: newUnstructured("cluster", "namespace", "", nil), desired: "cluster/namespace/"},
```

Comment:
I'd delete this test case

Reply:
Or maybe we need to split into 1 function that's use for keys, and 1 for indexes. Because calling a key func with name="" isn't really a valid thing.